File: web/src/statistics.ts
import { ENDPOINTS } from "./types";
import { assignGroups, groupingLabel, normalizeGrouping } from "./grouping";
import type { GroupingInput } from "./grouping";
import { boundedMinimum, erfc, numpySum } from "./numeric";
import type {
  Curve,
  CoxStatus,
  GroupingSpec,
  RiskPoint,
  Endpoint,
  EndpointResult,
  GeneData,
  SurvivalAnalysis,
} from "./types";

const MONTH_DAYS = 30.4375;

function finite(value: number | null): value is number {
  return value !== null && Number.isFinite(value);
}

export function kaplanMeier(time: number[], event: number[]): Curve {
  const order = time.map((_, index) => index).sort((a, b) => time[a] - time[b]);
  const orderedTime = order.map((index) => time[index]);
  const orderedEvent = order.map((index) => event[index]);
  const eventTimes = Array.from(
    new Set(orderedTime.filter((_, index) => orderedEvent[index] === 1)),
  ).sort((a, b) => a - b);
  const x = [0];
  const y = [1];
  let survival = 1;
  for (const eventTime of eventTimes) {
    const atRisk = orderedTime.filter((value) => value >= eventTime).length;
    const events = orderedTime.filter(
      (value, index) => value === eventTime && orderedEvent[index] === 1,
    ).length;
    if (atRisk <= 0) continue;
    const updated = survival * (1 - events / atRisk);
    x.push(eventTime / MONTH_DAYS, eventTime / MONTH_DAYS);
    y.push(survival, updated);
    survival = updated;
  }
  if (orderedTime.length) {
    x.push(Math.max(...orderedTime) / MONTH_DAYS);
    y.push(survival);
  }
  return {
    xMonths: x,
    survival: y,
    timeline: kmTimeline(time, event),
    n: time.length,
    events: event.reduce((sum, value) => sum + value, 0),
  };
}

export function logrank(
  time: number[],
  event: number[],
  high: boolean[],
): [number, number] {
  if (!high.some(Boolean) || high.every(Boolean))
    return [Number.NaN, Number.NaN];
  const eventTimes = Array.from(
    new Set(time.filter((_, index) => event[index] === 1)),
  ).sort((a, b) => a - b);
  let observedHigh = 0;
  let expectedHigh = 0;
  let variance = 0;
  for (const eventTime of eventTimes) {
    const risk = time.map((value) => value >= eventTime);
    const deaths = time.map(
      (value, index) => value === eventTime && event[index] === 1,
    );
    const n = risk.filter(Boolean).length;
    const nHigh = risk.filter((value, index) => value && high[index]).length;
    const d = deaths.filter(Boolean).length;
    const dHigh = deaths.filter((value, index) => value && high[index]).length;
    observedHigh += dHigh;
    expectedHigh += (d * nHigh) / n;
    if (n > 1) {
      variance += (d * (n - d) * nHigh * (n - nHigh)) / (n * n * (n - 1));
    }
  }
  if (!(variance > 0)) return [Number.NaN, Number.NaN];
  const chi2 = (observedHigh - expectedHigh) ** 2 / variance;
  return [chi2, erfc(Math.sqrt(chi2 / 2))];
}

export const COX_MESSAGES: Record<Exclude<CoxStatus, "ok">, string> = {
  empty_group: "The comparison leaves one expression group empty.",
  no_events: "No events were observed; a hazard ratio cannot be estimated.",
  no_information:
    "The groups have no overlapping event risk sets; the hazard ratio is unavailable.",
  separation:
    "The groups are completely separated at event times; no finite hazard ratio exists.",
  not_converged:
    "The Cox model did not converge; the hazard ratio is unavailable.",
};

export function coxBinary(
  time: number[],
  event: number[],
  high: boolean[],
): [number, number] {
  const { hr, p } = coxFit(time, event, high);
  return [hr, p];
}

export function coxFit(time: number[], event: number[], high: boolean[]) {
  const unavailable = (status: CoxStatus) => ({
    hr: Number.NaN,
    p: Number.NaN,
    status,
  });
  if (!high.some(Boolean) || high.every(Boolean))
    return unavailable("empty_group");
  if (!event.some((x) => x === 1)) return unavailable("no_events");
  const eventTimes = [...new Set(time.filter((_, i) => event[i] === 1))].sort(
    (a, b) => a - b,
  );
  let mixedLow = 0,
    mixedHigh = 0;
  const sets = eventTimes.map((t) => {
    const risk = time.flatMap((x, i) => (x >= t ? [i] : []));
    const deaths = time.flatMap((x, i) =>
      x === t && event[i] === 1 ? [i] : [],
    );
    const nh = risk.filter((i) => high[i]).length,
      nl = risk.length - nh;
    const dh = deaths.filter((i) => high[i]).length,
      d = deaths.length;
    if (nl && nh) {
      mixedHigh += dh;
      mixedLow += d - dh;
    }
    return { risk, nh, nl, dh, d };
  });
  if (mixedHigh + mixedLow === 0) return unavailable("no_information");
  if (!mixedHigh || !mixedLow) return unavailable("separation");
  const score = (beta: number) => {
    const w = Math.exp(beta);
    return sets.reduce(
      (sum, { nl, nh, d, dh }) => sum + dh - (d * nh * w) / (nl + nh * w),
      0,
    );
  };
  let bound = 8;
  while (!(score(-bound) > 0 && score(bound) < 0)) {
    bound *= 2;
    if (bound > 64) return unavailable("not_converged");
  }
  const objective = (beta: number) => {
    const bx = high.map((x) => beta * Number(x));
    let total = numpySum(bx.filter((_, i) => event[i] === 1));
    for (const { risk, d } of sets) {
      total -= d * Math.log(numpySum(risk.map((i) => Math.exp(bx[i]))));
    }
    return -total;
  };
  const beta = boundedMinimum(objective, -bound, bound);
  if (!Number.isFinite(beta)) return unavailable("not_converged");
  let hessian = 0;
  for (const { risk, d } of sets) {
    const weights = risk.map((i) => Math.exp(beta * Number(high[i])));
    const sum = numpySum(weights);
    const weighted = numpySum(weights.map((w, i) => w * Number(high[risk[i]])));
    const mean = weighted / sum;
    hessian -= d * (weighted / sum - mean * mean);
  }
  if (!(hessian < 0) || !Number.isFinite(hessian))
    return unavailable("not_converged");
  const z = beta / Math.sqrt(-1 / hessian);
  return {
    hr: Math.exp(beta),
    p: erfc(Math.abs(z) / Math.SQRT2),
    status: "ok" as CoxStatus,
  };
}

export function kmTimeline(time: number[], event: number[]): RiskPoint[] {
  let survival = 1,
    greenwood = 0;
  return [...new Set(time)]
    .sort((a, b) => a - b)
    .map((at) => {
      const n = time.filter((x) => x >= at).length;
      const d = time.filter((x, i) => x === at && event[i] === 1).length;
      const censored = time.filter((x, i) => x === at && event[i] === 0).length;
      if (d) {
        survival *= 1 - d / n;
        greenwood = n > d ? greenwood + d / (n * (n - d)) : Infinity;
      }
      return {
        timeMonths: at / MONTH_DAYS,
        survival,
        atRisk: n,
        events: d,
        censored,
        greenwood: Number.isFinite(greenwood) ? greenwood : null,
      };
    });
}

export function confidenceBounds(
  survival: number,
  greenwood: number | null,
  level = 0.95,
): [number, number] {
  const zs: Record<number, number> = {
    0.9: 1.6448536269514722,
    0.95: 1.959963984540054,
    0.99: 2.5758293035489004,
  };
  if (!zs[level])
    throw new Error("Confidence level must be 0.90, 0.95, or 0.99.");
  if (greenwood === null || !Number.isFinite(greenwood) || survival <= 0)
    return [NaN, NaN];
  const margin = zs[level] * Math.sqrt(greenwood);
  return [
    survival * Math.exp(-margin),
    Math.min(1, survival * Math.exp(margin)),
  ];
}

export function atRisk(curve: Curve, months: number): number {
  if (months <= 0) return curve.n;
  return (
    curve.timeline.find((point) => point.timeMonths >= months)?.atRisk ?? 0
  );
}

export function bhFdr(values: number[]): number[] {
  const output = values.map(() => Number.NaN);
  const valid = values
    .map((value, index) => ({ value, index }))
    .filter(({ value }) => Number.isFinite(value) && value >= 0 && value <= 1)
    .sort((a, b) => a.value - b.value);
  let running = 1;
  for (let rank = valid.length - 1; rank >= 0; rank -= 1) {
    running = Math.min(
      running,
      (valid[rank].value * valid.length) / (rank + 1),
    );
    output[valid[rank].index] = Math.max(0, Math.min(1, running));
  }
  return output;
}

export function prepareEndpoint(
  data: GeneData,
  endpoint: Endpoint,
  spec: GroupingSpec,
) {
  const clinical = data.clinical.endpoints[endpoint];
  const tpm = Array.from(data.expression, (encoded) =>
    encoded === data.missing ? NaN : 2 ** (encoded / data.scale) - 1,
  );
  const eligible = tpm.flatMap((value, i) =>
    Number.isFinite(value) &&
    finite(clinical.time[i]) &&
    (clinical.event[i] === 0 || clinical.event[i] === 1) &&
    clinical.time[i]! > 0
      ? [i]
      : [],
  );
  const assignment = assignGroups(
    eligible.map((i) => tpm[i]),
    eligible,
    data.gene.medians[endpoint],
    spec,
  );
  const included = eligible.flatMap((_, i) =>
    assignment.low[i] || assignment.high[i] ? [i] : [],
  );
  const indices = included.map((i) => eligible[i]);
  return {
    clinical,
    indices,
    time: indices.map((i) => clinical.time[i]!),
    event: indices.map((i) => clinical.event[i]!),
    high: included.map((i) => assignment.high[i]),
    eligibleN: eligible.length,
    excludedMiddle: eligible.length - included.length,
    lower: assignment.lower,
    upper: assignment.upper,
  };
}

function analyzeEndpoint(
  data: GeneData,
  endpoint: Endpoint,
  spec: GroupingSpec,
): EndpointResult {
  const {
    clinical,
    time,
    event,
    high,
    eligibleN,
    excludedMiddle,
    lower,
    upper,
  } = prepareEndpoint(data, endpoint, spec);
  const lowTime = time.filter((_, i) => !high[i]),
    lowEvent = event.filter((_, i) => !high[i]);
  const highTime = time.filter((_, i) => high[i]),
    highEvent = event.filter((_, i) => high[i]);
  const [chi2, p] = logrank(time, event, high);
  const fit = coxFit(time, event, high);
  return {
    endpoint,
    quality: clinical.quality,
    qualityNote: clinical.quality_note,
    n: time.length,
    nLow: lowTime.length,
    nHigh: highTime.length,
    events: event.reduce((sum, x) => sum + x, 0),
    eventsLow: lowEvent.reduce((sum, x) => sum + x, 0),
    eventsHigh: highEvent.reduce((sum, x) => sum + x, 0),
    cutoffTpm: lower === upper ? lower : NaN,
    lowerThreshold: lower,
    upperThreshold: upper,
    eligibleN,
    excludedMiddle,
    coxStatus: fit.status,
    logrankChi2: chi2,
    logrankP: p,
    logrankQ: NaN,
    coxHr: fit.hr,
    coxP: fit.p,
    low: kaplanMeier(lowTime, lowEvent),
    high: kaplanMeier(highTime, highEvent),
    warning: !time.length
      ? "No endpoint-valid samples."
      : fit.status === "ok"
        ? ""
        : COX_MESSAGES[fit.status],
  };
}

export function analyzeGeneData(
  data: GeneData,
  grouping: GroupingInput = "median",
): SurvivalAnalysis {
  const spec = normalizeGrouping(grouping);
  const endpointResults = ENDPOINTS.map((endpoint) =>
    analyzeEndpoint(data, endpoint, spec),
  );
  const adjusted = bhFdr(endpointResults.map((result) => result.logrankP));
  endpointResults.forEach((result, i) => {
    result.logrankQ = adjusted[i];
  });
  return {
    gene: data.gene.symbol,
    ensembl: data.gene.ensembl,
    cohort: data.cohort,
    cohortLabel: data.cohortLabel,
    sourceExpression: data.sourceExpression ?? "GDC STAR TPM",
    sourceSurvival: data.sourceSurvival ?? "PanCanAtlas TCGA-CDR",
    dataVersion: data.dataVersion,
    cutoff:
      spec.kind === "median"
        ? "median"
        : spec.kind === "tpm"
          ? spec.threshold
          : null,
    grouping: spec,
    groupingLabel: groupingLabel(spec),
    statisticsVersion: "1",
    endpoints: Object.fromEntries(
      endpointResults.map((result) => [result.endpoint, result]),
    ) as SurvivalAnalysis["endpoints"],
  };
}

export function formatP(value: number): string {
  if (!Number.isFinite(value)) return "NA";
  if (value < 0.001) return value.toExponential(1);
  if (value < 0.01) return value.toFixed(4).replace(/0+$/, "");
  if (value < 0.1) return value.toFixed(3).replace(/0+$/, "");
  return value.toFixed(2);
}

File: src/survscope/statistics.py
"""Small, auditable survival-statistics implementation."""

from __future__ import annotations

import math

import numpy as np
from scipy import optimize, stats

from .constants import MONTH_DAYS

COX_MESSAGES = {
    "empty_group": "The comparison leaves one expression group empty.",
    "no_events": "No events were observed; a hazard ratio cannot be estimated.",
    "no_information": (
        "The groups have no overlapping event risk sets; the hazard ratio is unavailable."
    ),
    "separation": (
        "The groups are completely separated at event times; no finite hazard ratio exists."
    ),
    "not_converged": "The Cox model did not converge; the hazard ratio is unavailable.",
}


def kaplan_meier(time: np.ndarray, event: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
    """Return post-step Kaplan-Meier coordinates in the input time units."""
    time = np.asarray(time, dtype=float)
    event = np.asarray(event, dtype=int)
    order = np.argsort(time, kind="stable")
    time = time[order]
    event = event[order]
    xs = [0.0]
    ys = [1.0]
    survival = 1.0
    for event_time in np.unique(time[event == 1]):
        at_risk = int(np.sum(time >= event_time))
        events = int(np.sum((time == event_time) & (event == 1)))
        if at_risk <= 0:
            continue
        updated = survival * (1.0 - events / at_risk)
        xs.extend([float(event_time), float(event_time)])
        ys.extend([survival, updated])
        survival = updated
    if len(time):
        xs.append(float(np.max(time)))
        ys.append(survival)
    return np.asarray(xs), np.asarray(ys)


def logrank_test(
    time: np.ndarray,
    event: np.ndarray,
    high: np.ndarray,
) -> tuple[float, float]:
    """Two-sided one-degree-of-freedom log-rank test."""
    time = np.asarray(time, dtype=float)
    event = np.asarray(event, dtype=int)
    high = np.asarray(high, dtype=int)
    if len(time) == 0 or len(np.unique(high)) < 2:
        return np.nan, np.nan

    observed_high = 0.0
    expected_high = 0.0
    variance = 0.0
    for event_time in np.unique(time[event == 1]):
        risk = time >= event_time
        deaths_at_time = (time == event_time) & (event == 1)
        n = int(np.sum(risk))
        n_high = int(np.sum(risk & (high == 1)))
        deaths = int(np.sum(deaths_at_time))
        deaths_high = int(np.sum(deaths_at_time & (high == 1)))
        if n <= 0:
            continue
        observed_high += deaths_high
        expected_high += deaths * n_high / n
        if n > 1:
            variance += deaths * (n - deaths) * n_high * (n - n_high) / (n * n * (n - 1))
    if variance <= 0 or not np.isfinite(variance):
        return np.nan, np.nan
    chi2 = (observed_high - expected_high) ** 2 / variance
    return float(chi2), float(stats.chi2.sf(chi2, 1))


def cox_binary(
    time: np.ndarray,
    event: np.ndarray,
    high: np.ndarray,
) -> tuple[float, float]:
    """Binary Cox proportional-hazards estimate using Breslow ties."""
    hr, pvalue, _ = cox_fit(time, event, high)
    return hr, pvalue


def cox_fit(
    time: np.ndarray,
    event: np.ndarray,
    high: np.ndarray,
) -> tuple[float, float, str]:
    """Legacy Breslow estimate with explicit identifiability and convergence checks."""
    time = np.asarray(time, dtype=float)
    event = np.asarray(event, dtype=int)
    x = np.asarray(high, dtype=float)
    if len(time) == 0 or len(np.unique(x)) < 2:
        return np.nan, np.nan, "empty_group"
    if int(event.sum()) == 0:
        return np.nan, np.nan, "no_events"
    event_times = np.unique(time[event == 1])
    risk_sets = []
    mixed_low = mixed_high = 0
    for event_time in event_times:
        risk = time >= event_time
        deaths = (time == event_time) & (event == 1)
        n_high = int(np.sum(risk & (x == 1)))
        n_low = int(np.sum(risk & (x == 0)))
        d_high = int(np.sum(deaths & (x == 1)))
        d = int(np.sum(deaths))
        risk_sets.append((n_low, n_high, d, d_high))
        if n_low and n_high:
            mixed_low += d - d_high
            mixed_high += d_high
    # Limiting scores at beta = +/- infinity: this also detects separation
    # when BOTH groups have events, but their event risk sets are separated.
    if mixed_low + mixed_high == 0:
        return np.nan, np.nan, "no_information"
    if mixed_low == 0 or mixed_high == 0:
        return np.nan, np.nan, "separation"

    def score(beta: float) -> float:
        weight = math.exp(beta)
        return sum(dh - d * nh * weight / (nl + nh * weight) for nl, nh, d, dh in risk_sets)

    bound = 8
    while not (score(-bound) > 0 and score(bound) < 0):
        bound *= 2
        if bound > 64:
            return np.nan, np.nan, "not_converged"

    def neg_loglik(beta: float) -> float:
        bx = beta * x
        total = float(np.sum(bx[event == 1]))
        for event_time in event_times:
            deaths = int(np.sum((time == event_time) & (event == 1)))
            risk_sum = float(np.sum(np.exp(bx[time >= event_time])))
            if risk_sum <= 0:
                return np.inf
            total -= deaths * math.log(risk_sum)
        return -total

    fit = optimize.minimize_scalar(
        neg_loglik,
        bounds=(-bound, bound),
        method="bounded",
        options={"xatol": 1e-5, "maxiter": 500},
    )
    if not fit.success or not np.isfinite(fit.x):
        return np.nan, np.nan, "not_converged"
    beta = float(fit.x)
    bx = beta * x
    hessian = 0.0
    for event_time in event_times:
        deaths = int(np.sum((time == event_time) & (event == 1)))
        risk = time >= event_time
        weights = np.exp(bx[risk])
        weight_sum = float(np.sum(weights))
        if weight_sum <= 0:
            return np.nan, np.nan, "not_converged"
        weighted_mean = float(np.sum(weights * x[risk]) / weight_sum)
        weighted_second = float(np.sum(weights * x[risk] * x[risk]) / weight_sum)
        hessian -= deaths * (weighted_second - weighted_mean * weighted_mean)
    if hessian >= 0 or not np.isfinite(hessian):
        return np.nan, np.nan, "not_converged"
    standard_error = math.sqrt(-1.0 / hessian)
    z_score = beta / standard_error
    return math.exp(beta), float(2 * stats.norm.sf(abs(z_score))), "ok"


def km_timeline(time: np.ndarray, event: np.ndarray) -> list[dict]:
    """Aggregate KM risk sets, including censor-only times; no subject records."""
    time = np.asarray(time, dtype=float)
    event = np.asarray(event, dtype=int)
    survival = 1.0
    greenwood = 0.0
    points = []
    for at in np.unique(time):
        n = int(np.sum(time >= at))
        d = int(np.sum((time == at) & (event == 1)))
        censored = int(np.sum((time == at) & (event == 0)))
        if d:
            survival *= 1.0 - d / n
            greenwood = greenwood + d / (n * (n - d)) if n > d else np.inf
        points.append(
            {
                "timeMonths": float(at / MONTH_DAYS),
                "survival": survival,
                "atRisk": n,
                "events": d,
                "censored": censored,
                "greenwood": float(greenwood) if np.isfinite(greenwood) else None,
            }
        )
    return points


def confidence_bounds(
    survival: float,
    greenwood: float | None,
    level: float = 0.95,
) -> tuple[float, float]:
    """Pointwise Greenwood/log intervals, matching R survfit(conf.type='log')."""
    zvalues = {0.90: 1.6448536269514722, 0.95: 1.959963984540054, 0.99: 2.5758293035489004}
    if level not in zvalues:
        raise ValueError("Confidence level must be 0.90, 0.95, or 0.99.")
    if greenwood is None or not np.isfinite(greenwood) or survival <= 0:
        return np.nan, np.nan
    margin = zvalues[level] * math.sqrt(greenwood)
    return survival * math.exp(-margin), min(1.0, survival * math.exp(margin))


def bh_fdr(values: list[float]) -> list[float]:
    """Benjamini-Hochberg adjustment, ignoring non-finite values."""
    output = np.full(len(values), np.nan, dtype=float)
    valid_indices = [
        index for index, value in enumerate(values) if np.isfinite(value) and 0 <= value <= 1
    ]
    if not valid_indices:
        return output.tolist()
    pvalues = np.asarray([values[index] for index in valid_indices], dtype=float)
    order = np.argsort(pvalues, kind="stable")
    ranked = pvalues[order]
    count = len(ranked)
    adjusted = ranked * count / np.arange(1, count + 1)
    adjusted = np.minimum.accumulate(adjusted[::-1])[::-1]
    adjusted = np.clip(adjusted, 0, 1)
    restored = np.empty_like(adjusted)
    restored[order] = adjusted
    for source_index, value in zip(valid_indices, restored, strict=True):
        output[source_index] = value
    return output.tolist()


def format_p(value: float) -> str:
    if not np.isfinite(value):
        return "NA"
    if value < 0.001:
        return f"{value:.1e}"
    if value < 0.01:
        return f"{value:.4f}".rstrip("0")
    if value < 0.1:
        return f"{value:.3f}".rstrip("0")
    return f"{value:.2f}"
